File: c/Queue/queue.c
#include <stdio.h>
#include "queue.h"

void createQueue(Queue *queue)
{
    queue->back = MAX_SIZE - 1;
    queue->front = 0;
    queue->size = 0;
}

int isQueueEmpty(Queue *queue)
{
    return (queue->size == 0);
}

int isQueueFull(Queue *queue)
{
    return (queue->size == MAX_SIZE);
}

void enqueue(Queue *queue, QueueType item)
{
    queue->back = (queue->back + 1) % MAX_SIZE;
    queue->items[queue->back] = item;
    queue->size++;
}

QueueType dequeue(Queue *queue)
{
    QueueType item;
    queue->size--;
    item = queue->items[queue->front];
    queue->front = (queue->front + 1) % MAX_SIZE;
    return item;
}

QueueType front(Queue *queue)
{
    return queue->items[queue->front];
}

int queueSize(Queue *queue)
{
    return queue->size;
}

File: Java/Queue.java
package Java;

public class Queue<Type> {
    private final int MAX_SIZE = 100;
    private int front;
    private int back;
    private int size;
    private Type[] data;

    @SuppressWarnings("unchecked")
    public Queue() {
        this.front = 0;
        this.back = MAX_SIZE - 1;
        this.size = 0;
        this.data = (Type[]) new Object[MAX_SIZE];
    }

    public boolean isEmpty() {
        return (this.size == 0);
    }

    public boolean isFull() {
        return (this.size == MAX_SIZE);
    }

    public boolean enqueue(Type value) {
        if (this.isFull())
            throw new IllegalStateException("Queue overflow");
        this.data[(++this.back) % MAX_SIZE] = value;
        this.size++;
        return true;
    }

    public Type dequeue() {
        if (this.isEmpty())
            throw new IllegalStateException("Queue underflow");
        this.size--;
        Type item = this.data[this.front];
        this.front = ++this.front % MAX_SIZE;
        return item;
    }

    public Type front() {
        if (this.isEmpty())
            throw new IllegalStateException("Queue underflow");
        return this.data[this.front];
    }

    public int size() {
        return this.size;
    }
}
